File: src/runtime/companion-preludes/TacPrelude.cs
// Tachyon editor prelude for C# Tac companions.
//
// Never compiled or bundled by Tachyon (only `tac.cs` beside a `tac.html` is
// a companion). Copy this file next to your companion so IDE tooling
// resolves the implicit prelude symbols.
//
// ponytail: a real C# compiler still rejects Tac dialect forms such as bare
// FetchAsync(...) and `{}` query arguments; this stub only fixes symbol
// resolution. C# companions are a Tac dialect, not real C# — see the
// portable companion subset reference in the Tachyon README.

using System;

public class Tac {
    public bool Publish(string name, object value = null) => false;
    public object Env(string key, object fallback = null) => fallback;
    public void Rerender() {}
}

[AttributeUsage(AttributeTargets.All)] public class PublishAttribute : Attribute { public PublishAttribute(string name = "") {} }
[AttributeUsage(AttributeTargets.All)] public class SubscribeAttribute : Attribute { public SubscribeAttribute(string name = "") {} }
[AttributeUsage(AttributeTargets.All)] public class OnMountAttribute : Attribute {}

public static class LocalStorage {
    public static string GetItem(string key, string fallback = "") => fallback;
    public static void SetItem(string key, object value) {}
    public static void RemoveItem(string key) {}
}

public static class SessionStorage {
    public static string GetItem(string key, string fallback = "") => fallback;
    public static void SetItem(string key, object value) {}
    public static void RemoveItem(string key) {}
}

public static class Navigator {
    public static string Language() => "";
    public static bool IsOnline() => false;
}

public static class Location {
    public static string Href() => "";
    public static string Origin() => "";
}

public class FyloCollection {
    public object Find(object query) => null;
    public object Get(object id) => null;
    public object Create(object document = null) => null;
    public object Put(object document) => null;
    public object Delete(object id) => null;
}

public static class Fylo {
    public static FyloCollection Collection(string name) => new FyloCollection();
}

public class AppInfo { public string name = ""; public string version = ""; }

public static class App {
    public static bool IsAvailable() => false;
    public static AppInfo InfoAsync() => new AppInfo();
}

public static class Clipboard {
    public static string GetTextAsync() => "";
    public static void SetTextAsync(string text) {}
}

public static class FileSystem {
    public static string ReadTextAsync(string path) => "";
    public static void WriteTextAsync(string path, string text) {}
    public static object ReadDirAsync(string path) => null;
    public static object PathsAsync() => null;
}

public static class Shell {
    public static object ExecAsync(string command, string[] args = null, string cwd = null) => null;
}

public static class Browser {
    public static void OpenAsync(string url) {}
}

public static class Share {
    public static void TextAsync(string text, string title = null) {}
}

public static class Haptics {
    public static void ImpactAsync() {}
}

public static class FilePicker {
    public static string OpenTextAsync() => "";
    public static void SaveTextAsync(string name, string text) {}
}

public static class Capabilities {
    public static bool Supports(string capability) => false;
    public static string StateAsync(string capability) => "unsupported";
}

public static class Secrets { public static string GetAsync(string key) => null; public static void SetAsync(string key, string value) {} public static void DeleteAsync(string key) {} }
public static class Auth { public static object VerifyUserAsync(string reason) => null; }
public static class Geolocation { public static object CurrentAsync(object options = null) => null; }
public static class Notifications { public static void ShowAsync(string title, object options = null) {} }
public static class Media { public static object GetUserMediaAsync(object constraints) => null; }

File: src/runtime/companion-preludes/TacPrelude.kt
// Tachyon editor prelude for Kotlin Tac companions.
//
// Never compiled or bundled by Tachyon (only `tac.kt` beside a `tac.html` is a
// companion). Copy this file next to your companion so editors and `kotlinc`
// resolve the implicit prelude. A Kotlin companion that compiles with this
// prelude under real `kotlinc` is valid Tac Kotlin — the CI parse check in
// tests/compiler/companion-real-language.test.js relies on that property.

annotation class publish(val name: String = "")
annotation class subscribe(val name: String = "")
annotation class onMount

open class Tac {
    fun publish(name: String, value: Any? = null): Boolean = false
    fun env(key: String, fallback: Any? = null): Any? = fallback
    fun rerender() {}
}

object localStorage {
    fun getItem(key: String, fallback: String = ""): String = fallback
    fun setItem(key: String, value: Any?) {}
    fun removeItem(key: String) {}
}

object sessionStorage {
    fun getItem(key: String, fallback: String = ""): String = fallback
    fun setItem(key: String, value: Any?) {}
    fun removeItem(key: String) {}
}

object navigator {
    fun language(): String = ""
    fun isOnline(): Boolean = false
}

object location {
    fun href(): String = ""
    fun origin(): String = ""
}

fun fetch(url: String, init: Any? = null): Any? = null

class FyloCollection {
    fun find(query: Any?): Any? = null
    fun get(id: Any?): Any? = null
    fun create(document: Any? = null): Any? = null
    fun put(document: Any?): Any? = null
    fun patch(id: Any?, patch: Any? = null): Any? = null
    fun delete(id: Any?): Any? = null
    fun list(): Any? = null
}

object fylo {
    fun collection(name: String): FyloCollection = FyloCollection()
}

class AppInfo {
    val name: String = ""
    val version: String = ""
}

object app {
    fun isAvailable(): Boolean = false
    fun info(): AppInfo = AppInfo()
}

object clipboard {
    fun readText(): String = ""
    fun writeText(text: String) {}
}

object fileSystem {
    fun readText(path: String): String = ""
    fun writeText(path: String, text: String) {}
    fun readDir(path: String): Any? = null
    fun stat(path: String): Any? = null
    fun mkdir(path: String): Any? = null
    fun remove(path: String): Any? = null
    fun paths(): Any? = null
}

object shell {
    fun exec(command: String, args: Any? = null, cwd: String? = null): Any? = null
}

object browser {
    fun open(url: String) {}
}

object share {
    fun text(text: String, title: String? = null) {}
}

object haptics {
    fun impact() {}
}

object filePicker {
    fun openText(): String = ""
    fun saveText(name: String, text: String) {}
}

object capabilities {
    fun supports(capability: String): Boolean = false
    fun state(capability: String): String = "unsupported"
}

object secrets { fun get(key: String): String? = null; fun set(key: String, value: String) {}; fun delete(key: String) {} }
object auth { fun verifyUser(reason: String): Any? = null }
object geolocation { fun current(options: Any? = null): Any? = null }
object notifications { fun show(title: String, options: Any? = null) {} }
object media { fun getUserMedia(constraints: Any?): Any? = null }
